### PR TITLE
URL API

### PR DESCRIPTION
**Pull request by bagder**
The [URL API](https://github.com/curl/curl/wiki/URL-API). Man pages with all the details and a fair amount of tests are included. I encourage all interested parties to try it out and report your experiences, both good and bad!

<hr>

I'm saving these items until after the initial merge and done as separate PRs:

- make libcurl use this parser internally too to avoid different treatment
- add `CURLOPT_CURLU` to allow passing in a URL handle to work with
- [CURLU_CANONICAL](https://curl.haxx.se/mail/lib-2018-08/0151.html) - if there's a desire.

### CODE REVIEW

**Comment by xquery** [comment]
looks good to me, there is some good inspiration over at  https://github.com/uriparser/uriparser/blob/master/test/test.cpp 
for other test variants.

File: tests/unit/Makefile.inc [hunk 1]
```diff
@@ -95,3 +95,4 @@ unit1608_CPPFLAGS = $(AM_CPPFLAGS)
 
 unit1609_SOURCES = unit1609.c $(UNITFILES)
 unit1609_CPPFLAGS = $(AM_CPPFLAGS)
+
```
Comment (vszakats) [inline tests/unit/Makefile.inc]: Minor nit: An unnecessary empty line got added here.

Comment (vszakats) [inline tests/unit/Makefile.inc]: Minor nit: An unnecessary empty line slipped in here.

Comment (bagder) [reply tests/unit/Makefile.inc]: oops

**Comment by vszakats** [comment]
Is there any reason why `const` isn't used in this prototype for `URL`?:
```c
CURLUcode curl_url(char *URL, CURLURL **urlhandle, unsigned int flags)
```
This function makes a call to `parseurl()`, where a const could added as well, and where internally the buffer isn't written nor passed to functions that write into it. ...it's possible I missed something reading the code.

**Comment by bagder** [comment]
No reason at all really. It's just an old weakness of mine to skip out on the 'const's... I'll fix. Thanks!

**Comment by bagder** [comment]
More minor fixes, squashed and rebased.

File: docs/libcurl/curl_url.3 [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+.\" *
+.\" **************************************************************************
+.TH curl_url 3 "6 Aug 2018" "libcurl" "libcurl Manual"
+.SH NAME
+curl_url - parse a URL and create a CURLURL handle
+.SH SYNOPSIS
+.B #include <curl/curl.h>
+
+.nf
+CURLUcode curl_url(char *url,
```
Comment (vszakats) [inline docs/libcurl/curl_url.3]: Thanks for the constifying, it's looking good! One minor: `curl_url()` prototype above needs to be synced with this change.

**Comment by bagder** [comment]
I think coveralls got some hiccups there when it claims a -3.4% less test coverage, as this patch provides quite a bit of testing already. I'm ignoring that for now, as I hope to add more tests anyway as I go along and fix the remaining parts.

**Comment by bagder** [comment]
I'm not entirely happy with how `curl_url()` works just now, as just now we have two ways we can set a URL to the handle. I'm switching from:

# Current
~~~c
  CURLURL *h;
  rc = curl_url("http://example.com", &h, 0); /* create handle *
~~~

I suspect that way of storing the handle pointer is going to cause users grief. A bit too unusual.

# New
~~~c
  CURLURL *h = curl_url(); /* create handle */
  curl_url_set(h, CURLUPART_URL, "http://example.com", 0);
~~~

It is more in style with how our other APIs work and makes sure there's one single way to set URL in the handle,

# Syntax-check a URL

A downside with this approach is that there's no longer any "shortcut" for just syntax-checking a URL without creating a handle, but then again I'm not convinced that's a very important feature.

  rc = curl_url("http://example.com", &h, 0); /* create handle *

**Comment by bagder** [comment]
I've synced the wiki page to match how the current patch implements the API.

**Pull request closed without merging**